python/ray/serve/tests/test_replica_scheduler.py

```diff
@@ -563,11 +670,78 @@ async def test_prefer_replica_on_same_node(pow_2_scheduler, fake_query):
     assert all(replica == r2 for replica in await asyncio.gather(*tasks))
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "pow_2_scheduler",
+    [{"prefer_local_node": True, "prefer_local_az": True, "az": SCHEDULER_AZ}],
+    indirect=True,
+)
+async def test_prefer_replica_in_same_az(pow_2_scheduler, fake_query):
+    """
+    Verify that the scheduler prefers
+    * replicas that are colocated on the same node
+    * then replicas that are colocated in the same AZ
+    * lastly fall back to all replicas
+    """
```

Comment:
please also add a test for where `prefer_local_az` is off and there are nodes in multiple AZs (should get spread across them)

this would've caught the bug I pointed out where you're ignoring the prefer_local_az flag

some other cases to add:

- parametrize this test for `prefer_local_az=True` and `prefer_local_node=False`
- add another test for `prefer_local_az=False` and `prefer_local_node=True`, verify that if we spill beyond the node it balances across AZs

Reply:
Also please test the case where we spill back after not being able to find matching multiplexed ID

Reply:
> Also please test the case where we spill back after not being able to find matching multiplexed ID

Per offline discussion, not testing this because multiplexing is a separate code path.

Added the other tests cases to `test_replica_scheduler.py`!